lib/storage/sfCacheSessionStorage.class.php

```diff
@@ -29,140 +29,164 @@ class sfCacheSessionStorage extends sfStorage
+                $this->options['session_cookie_domain'],
+                $this->options['session_cookie_secure'],
+                $this->options['session_cookie_httponly']
+            );
+
+            $this->data = [];
+        } else {
+            // load data from cache. Watch out for the default case.
+            $raw = $this->cache->get($this->id, null);
+            if (null === $raw) {
+                $this->data = [];
+            } else {
+                $data = json_decode($raw, true);
+                if (is_array($data)) {
+                    $this->data = $this->upgradeNestedRouteObjects($data);
+                } else {
+                    $data = @unserialize($raw);
+
+                    // We test 'b:0' special case, because such a string would result
+                    // in $data being === false, while raw is serialized
+                    // see http://stackoverflow.com/questions/1369936/check-to-see-if-a-string-is-serialized
+                    if ($raw === 'b:0;' || $data !== false) {
+                        $this->data = $data;
+                    } else {
+                        // Probably an old cached value (BC)
+                        $this->data = $raw;
+                    }
+                }
+            }
+
+            if (sfConfig::get('sf_logging_enabled')) {
+                $this->dispatcher->notify(new sfEvent($this, 'application.log', ['Restored previous session']));
+            }
         }
-      }
+        session_id($this->id);
+        $this->response->addCacheControlHttpHeader('private');
 
-      if(sfConfig::get('sf_logging_enabled'))
-      {
-        $this->dispatcher->notify(new sfEvent($this, 'application.log', array('Restored previous session')));
-      }
+        return true;
     }
-    session_id($this->id);
-    $this->response->addCacheControlHttpHeader('private');
 
-    return true;
-  }
+    /**
+     * Recursive method to search through a json_decode()'d cache payload for
+     * any values which _would have been_ sfRoute objects before serialisation.
+     * If any are found, then also rebuilds the sfRoute object from the cached
+     * values and injects it back into the array.
+     *
+     * @param  array  $raw  A dictionary of key => value pairs
+     */
+    protected function upgradeNestedRouteObjects(array $raw): array
+    {
+        foreach ($raw as $key => $value) {
+            if (is_array($value)) {
+                if (
+                    $key == 'route'
+                    && isset(
+                        $value['pattern'],
+                        $value['defaults'],
+                        $value['requirements'],
+                        $value['options']
+                    )
+                ) {
+                    $raw[$key] = sfRoute::jsonUnserialize($value);
+                    continue;
+                }
+
+                $raw[$key] = $this->upgradeNestedRouteObjects($value);
+            }
+        }
+
+        return $raw;
+    }
```

Comment:
When I was testing, I found the `sfRoute` object buried down at `$raw[sfUser::ATTRIBUTE_NAMESPACE][sfUser::ATTRIBUTE_NAMESPACE]['manager-url-options']['route']` (yes, it was nested twice under the same key, goodness only knows why).

I'm worried that `sfRoute` objects could exist at varying keys, so I've used generic recursion to find and replace any which are present, rather than coding for the one specific use case seen so far. There didn't seem to be any noticeable slowdown in execution during testing due to the overhead of parsing the content in this manner.

I'm also hoping that if we discover _other_ object classes which may be found in the session cache, we can add a similar JSON (de)serialisation strategy to those classes too.